V3

tlbdk opened the pull request:
This is fairly big rewrite to make it much easier to use, extend or use as base for other validation frameworks.

* Add more flexible and stable type interference to enable creating custom validators outside this package
* Added much better readme
* Added Sample validator
* Added Union validator
* Add support for earlyFail
* Simplify adding new validators by move more code to the base class
* Remove old function type (StringValue, Integer, etc)
* Clean up Fail type so string is generate in one place
* Add missing basic validators(Boolean, Null, etc.)

Commit:
Get of type mapping and legacy validator functions

Commit:
Upgrade dependencies

Commit:
Fix linting

Commit:
Add todo

Commit:
Start work on moving required into options

Commit:
Add initial structure for union

Commit:
Minor fixes for object

Commit:
Default optimize to false as we don't support having nested objects that gets optimize more than once.

Commit:
Avoid passing the optimized validate to the codegen

Commit:
Implement is functions for all validators

Commit:
Remember to bind to this

Commit:
Cleanup option handling

Commit:
Update deps

Commit:
Add initial union implementation

Commit:
Implement initial optimization for union

Commit:
Add todo

Commit:
Overwrite name when extending Error

Commit:
Fix issue with mixing optimized and not optimized validators
Simplify validation code

Commit:
Simplify default codegen

Commit:
Remove duplicate test

Commit:
Add unknown validator

Commit:
Add more tests and fix key name for failures

Commit:
Cleanup todo list

Commit:
Add support for micro seconds to DateTimeValidator

Commit:
Add more complex test examples

Commit:
Update TODO

Commit:
Add BooleanValidator
Add NullValidator

Commit:
Clean test namings

Commit:
Added sample validator

Commit:
Added string enum wrapper for union

Commit:
Added earlyFail

Commit:
Move todo to clubhouse

Commit:
Add DateTime detection to sample validator

Commit:
Add support for exporting validator init code with toString function

Commit:
Add initial type export

Commit:
Fix naming of error types

Commit:
Add optimization for common object union types

Commit:
Add check for required to union optimization

Commit:
Allow the benchmark to earlyFail

Commit:
Fix escape of key name

Commit:
Merge old union types into union

Commit:
Simplify context to string

Commit:
Move received fail message into Fail object

Commit:
Fix OutOfRange fails

Commit:
Create initial readme

Commit:
Add basic documentaion for each validator

Commit:
Add more documentation

Commit:
Add more code coverage

ofweb commented on README.md:
```diff
@@ -1,59 +1,170 @@
+* [Float](#floatvalidatormin-max): Validate that a value is a float within a min and max limit
+* [IntegerString](#integerstringvalidatormin-max): Validate that a value is a integer string within a min and max limit
+* [Integer](#integervalidatormin-max): Validate that a value is a integer within a min and max limit
+* [Null](#nullvalidator): Validate that a value is null
```
why not a validator for `undefined`?  Now that we have a union validator that might be useful.

tlbdk replied:
Will do that.

ofweb commented on README.md:
```diff
@@ -1,59 +1,170 @@
+Validation functions:
+
+* .validate(value) : Returns list of errors found in the validation
+* .isType(value, listOfErrors) : Type guard for the type
```
I assume that this does not run the validation again, but it might be nice to have that be written explicitly.

tlbdk replied:
Fixed

ofweb commented on README.md:
```diff
@@ -1,59 +1,170 @@
+* optimized: Create optimized function and use this for validation (default: false)
+* required: Is this validator required (default: true)
+* nullCheck: Enable or disable null check (default: true)
+* earlyFail: Return on first error (default: false)
```
very minor thing: but is this not normally called failFast?

tlbdk replied:
Might be, should I change it?

ofweb replied:
No, it is fine.

ofweb commented on README.md:
````diff
@@ -87,10 +198,397 @@ if(objectValidator.isValid(unknownValue)) {
   console.log(knownValue.positions.length()) // unknownValue.unitId has been type cast to array
 }
+```
+
````
It would be nice with an example of moving from a sample validator to a non sample one. I guess that it is done with `.toString()`, but an example would make it more explicit that it can be used as a stepping stone.

tlbdk replied:
Added.

ofweb commented on README.md:
```diff
@@ -87,10 +198,397 @@ if(objectValidator.isValid(unknownValue)) {
+
+Function: isUnknown
+
+The unknown validator only makes sense to use with another validate like the object validator to require a key even when the type is not known.
```
I do not understand how this is used and why

tlbdk replied:
fx. If I want to validate a union type but want to require a key but not validate it:

``` json5
{
  "type": "type1"
  "somekey": {. // I'm too lazy to validate this right now }
}
```

``` typescript
if (obj.type === 'type1') {
  // Now i'm validating obj.somekey
}
```

ofweb commented on README.md:
```diff
@@ -1,59 +1,170 @@
 # Object-validator
 
```
Maybe also mention that it has no dependencies

tlbdk replied:
Done

ofweb commented on README.md:
```diff
@@ -1,59 +1,170 @@
+
+Common options:
+
+* optimized: Create optimized function and use this for validation (default: false)
```
Any reason for `optimized` to not default to true?

tlbdk replied:
There used to be as you could not mix optimized and non-optimized validators, but this is fixed now

tlbdk replied:
Will make it default

ofweb commented on src/validators/datetime.ts:
```diff
@@ -1,106 +1,105 @@
+  return false
+}
+
 const dateTimePattern = /^([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])[Tt]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\.[0-9]+)?(([Zz])|([+|-]([01][0-9]|2[0-3]):[0-5][0-9]))$/
```
Don't RFC 3339 require 4 digit years?

tlbdk replied:
Replaced with this:

```
^([0-9]{4})...
```

ofweb commented on src/validators/datetime.ts:
```diff
@@ -1,106 +1,105 @@
+  return false
+}
+
 const dateTimePattern = /^([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])[Tt]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\.[0-9]+)?(([Zz])|([+|-]([01][0-9]|2[0-3]):[0-5][0-9]))$/
```
This doesn't ensure that it is a valid date eg `2018-02-31T13:37:00Z`
Could we try calling `Date.parse` if the regex matches?

tlbdk replied:
Add to todo list

ofweb commented on src/validators/string.ts:
```diff
@@ -1,46 +1,45 @@
 
-  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions, required = true) {
-    super()
+  public constructor(minLength = 0, maxLength = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
```
Now that we are doing a major bump, could we do something more general for the value encoded in string validators? 
We have `floatString`, `integerString` and `dateTime` it would be cool if string took a function as input to validate the content of the string.

ofweb replied:
It could also handle a regex as input to remove the explicit regex class

ofweb replied:
Maybe a conversion function and a validator for validating the converted value?

tlbdk replied:
This would make it quite a bit harder to do the codegen function so it easier to have the high level validators.

ofweb commented on src/validators/union.ts:
```diff
@@ -0,0 +1,403 @@
+  }
+}
+
+export class EnumValidator<T, O = never> extends UnionValidator<T, O> {
```
Is this needed or could it be an example?

tlbdk replied:
This is used to make TypeScript enums

ofweb commented on src/validators/union.ts:
```diff
@@ -0,0 +1,403 @@
+}
+
+export class DateTimeOrDateValidator<O = never> extends UnionValidator<string | Date | O> {
+  public constructor(options?: ValidatorOptions) {
```
I guess the only reason these (`DateTimeOrDateValidator`, `FloatOrFloatStringValidator` and `IntegerOrIntegerStringValidator`) are here is to handle the error message?

Is the normal error message from union that bad? or could we just drop them?

tlbdk replied:
backwards compatibility with the old error messages

ofweb commented on src/validators/unknown.test.ts:
```diff
@@ -0,0 +1,52 @@
+    describe('OptionalString', () => {
+      it('accepts empty value', () => {
+        const validator = new OptionalUnknown()
+        expect(validator.validate(undefined)).toStrictEqual([])
+        expect(validator.validate(undefined)).toStrictEqual([])
```
why are we testing the same thing twice?

tlbdk replied:
Removed

ofweb commented on src/validators/unknown.test.ts:
```diff
@@ -0,0 +1,52 @@
+    describe('RequiredUnknown', () => {
+      it('rejects empty value', () => {
+        const validator = new RequiredUnknown()
+        expect(validator.validate(null)).toStrictEqual([new RequiredFail('Is required', null)])
+        expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required', undefined)])
```
Is it not a bit arbitrary that null and undefined is not unknown?
The typescript type accepts both

tlbdk replied:
Yes and you could argument that null should also be a possibility with this implementation, the problem I'm trying to solve is that we want to have a key in object but don't want to validate it. 

fx. 

``` javascript
let obj = {
  "key1": undefined, // Would get excluded when exported to JSON
  "key2": null // would get exported
}

// But when looking a the parsed object I can't see if the key was missing or undefined inside the UnknownValidator
console.log(obj[key1] === undefined) // true
console.log(obj[key3] === undefined) // true

console.log(obj[key2] === undefined) // false

// I would have to replace the UnknownValidator in ObjectValidator when I see it
console.log("key1" in obj) // true
console.log("key3" in obj) // false
console.log(obj.hasOwnProperty("key1")) // true
console.log(obj.hasOwnProperty("key3")) // false
```

ofweb replied:
If we have a `requiredUnknown` and an `optionalUnknown` that does not accept `null` is that not the same as an `undefined` and `!undefined` validators?

I would prefer validators that compose well over handling special cases.

ofweb replied:
You could even go as far as having a `not` validator.

tlbdk replied:
Will remove null from Unknown

tlbdk replied:
They have now been split in to seperate validators:

* Required... : T
* Optional... : T | undefined
* Nullable... : T | null
* OptionalNullable... : T | null | undefined

ofweb commented on src/validators/union.ts:
```diff
@@ -0,0 +1,403 @@
+  const errors: ValidationFailure[] = []
+
+  if (!options?.every && schema.length > 1 && schema.every((v): v is ObjectValidator => v instanceof ObjectValidator)) {
+    const unionKey = findUnionKey(schema)
```
We could also handle the case where all validators are of different types eg. string or number.
That might also let us make the error message better.

tlbdk replied:
Add as future feature.

ofweb commented on src/validators/exact-string.ts:
```diff
@@ -1,90 +1,93 @@
   return []
 }
 
 export class ExactStringValidator<O = never> extends ValidatorBase<string | O> {
```
instead of limiting this to only string, what about an `ExcatValue` validator? that handles simple values.

tlbdk replied:
Add a todo for building a generic ExactValidator

ofweb commented on package.json:
```diff
@@ -35,30 +35,30 @@
   },
```
`npm run bencmark` does not work for me.

tlbdk replied:
Fixed

Commit:
Make optimize default to on

Commit:
Fix findings from review

Commit:
Fix benchmark script in npm run

Commit:
Added undefined validator

Commit:
Make it possible to use interface with object validators

Commit:
Fix docs

Commit:
Include UndefinedValidator in export

Commit:
Require 4 digit year for datetime

Commit:
Also change the optimize default for toString()

Commit:
UnknownValidator should only have special handling for undefined

Commit:
Refactor to use new type construction
Remove null from Optional...
Clean up tests for all validations
Make ...Validators abstract to make sure we have type information

Commit:
Only allow to set options that don't change types

Commit:
Add NullableBoolean and OptionalNullableBoolean

Commit:
Update documentation

Commit:
Add Nullable and OptionalNullable for all validators with better testing

Commit:
Improve code coverage

tlbdk commented:
Reimplement the type system in a more solid version and made much harder to use the validators incorrectly
